### DISCUSSION TITLE
Is there a performance issue when having over 1000 objects?

### DISCUSSION THREAD

**Discussion by riftek** [dyn4j/dyn4j#271] [Q&A]
When I have over 1000 objects, i see my cpu usage around 16% but theres lag between updates.
is it possible that its something in dyn4 ? I set most of the objects to setActive(false)

**Comment by wnbittle** [comment]
In my test I used `Bucket` from the samples project as a baseline and increased it from 200 to 1000 objects.  With or without enabled/active bodies I got around 16% CPU utilization.  If I pause the simulation, which stops calling `World.update(double)` entirely, the simulation stays around 14% utilization.  If I zoom in, which limits what is rendered, I see lower CPU utilization, somewhere around 11%.  If I pause and then move the screen so nothing is in view, I see 4% utilization.  If I remove _all_ rendering the CPU utilization is halved, hovering around 8%.

As far as lag between updates - this is because the `World.update(double)` will perform, at most, one update per call.  It will only perform an update if the first parameter is greater than or equal to `Settings.getStepFrequency()`.  If the first parameter is much greater, then the simulation can't catch up and the simulation seems 'slow' or 'lagged'.

Here are a few thoughts on how to get more performance outside of the obvious, which would be to reduce the number of bodies:
- Change the `Settings.setStepFrequency(double)` to 1/30 or even lower
- Try using `World.updatev(double)` instead - not recommended, but can work in some scenarios
- Try lowering the `Settings.setVelocityConstraintSolverIterations(int)` and `Settings.setPositionConstraintSolverIterations(int)` to 6 and 2 respectively (these are going to get lowered in the next minor release anyway)
- Avoid the use of curved shapes, `Circle`, `Ellipse`, `Slice`, etc. they are expensive for penetration detection

When I profiled the modified Bucket sample with VisualVM I saw ~60% CPU time in render and ~20% in dyn4j on average.

![image](https://user-images.githubusercontent.com/11467997/211680192-5775173a-5dfa-4c15-a3d5-2d8500557a1d.png)

**Reply by riftek** [reply]
Thanks, Ive tried the above, unfortunately there is still an issue, calling update takes 150ms to perform. they are all polygon fixtures.

**Comment by wnbittle** [comment]
Something seems off - In my test I'm seeing an average of `10` milliseconds for each `World.update(double)`. If I only adjust `Settings.setVelocityConstraintSolverIterations(int)` and `Settings.setPositionConstraintSolverIterations(int)` as described above I can get down to an average of `5` milliseconds per update.  

What's the host system look like?  If you are mobile GC pressure could be an issue.  There are so many things that could be wrong though.  I'd try to hook up to a profiler and pinpoint the issue.

![image](https://user-images.githubusercontent.com/11467997/211912352-cf317643-56f5-48e8-b8bd-92753ab6a58d.png)

![image](https://user-images.githubusercontent.com/11467997/211912499-4a2ae49c-659d-43bd-8a14-a9d381cfdec6.png)

**Reply by riftek** [reply]
I have a snapshot of jprofiler,
![image](https://user-images.githubusercontent.com/30284512/211979256-8b02de5d-3750-4cb5-b12a-25c61a2b929f.png)
its difficult to see with the profiler the time duration.
but I saw these in the beginning take up near the 150ms that I described.

Please let me know if there's anything else I can try. thanks for the help.

**Reply by riftek** [reply]
here is jprofiler for bucket sample.
![image](https://user-images.githubusercontent.com/30284512/211982783-14cf5fb6-db1f-4db9-8d67-e0decc38cbaf.png)
world.update is taking 50ms.. 
what CPU do you have?
mine is 3.5GHz on one thread

**Reply by wnbittle** [reply]
My hardware is pretty new (1 year old), but lower spec - 2.50GHz, one thread with Java 11 OpenJDK on Windows 10.0.19045 x64.

What's interesting from your profile is what is showing up as _slow_.  `SequentialImpulses.getMassCoefficient` is a super simple method, but somehow it's taking 433 ms?!?!  This method doesn't even show up in my profiling.  What version/VM are you using (OpenJDK, Oracle, etc.)?

What this profile tells me is that some methods aren't being optimized by the JIT compiler and that some methods aren't being inlined.  Maybe optimizations or the JIT is disabled?  It almost looks like you are in full interpretive mode or the VM has stopped optimizing by the time it reaches dyn4j code.  When I turn off the JIT compiler using `-Djava.compiler=NONE` I get similar performance as you - `230` ms per `World.update(double)` for the `Bucket` sample with 1000 objects.

**Reply by riftek** [reply]
The time is misleading , the logs from the profiler makes the application slower. It seems to inject code
I am using openJDK 11.

**Reply by wnbittle** [reply]
Visual VM has both a sampler and a profiler.  The profile does inject code and slow the application down considerably.  The sampler is much better at not doing this - not familiar with `jprofiler` and if it has a similar concept.

How are you running the sample?  Command line, IDE, etc.?  Maybe a VM argument is being set that shouldn't be?  I'm running out of ideas...

**Reply by wnbittle** [reply]
I had a couple of other thoughts (they are a long shot for sure) - 

What CPU do you have Intel or AMD or?  I've noticed in the past that AMD hardware (or at least the AMD hardware I had at the time) was significantly slower than Intel.

Also, how many cores?  I have an 8 core (16 logical) processor.  We're both single threaded and therefore utilizing a single CPU, but the JIT compiler runs in parallel and tailors optimization based on the host system.

Can you do a test on a different system to see if you get a different result?

**Reply by riftek** [reply]
I have 4 cores 8 logical processors, I do not have another machine to test on apart from a cloud server.
Ill message here when I come right. Thanks! its and intel i7 7th gen

**Reply by riftek** [reply]
im getting 15ms now, in bucket...Still investigating
